### ISSUE TITLE
r/folder: Tags being queried fetched when no tags specified, causing permission errors

### ISSUE THREAD

**Issue by vancluever** [terraform-providers/terraform-provider-vsphere#194]
Carrying this over from #191 as the issue seems to be a bug with fetching tags when they don't exist. cc @deberon

Current status: can't reproduce with this basic config and a user that has a role that only has folder permissions:

```hcl
data "vsphere_datacenter" "dc" {
  name = "dc1"
}

resource "vsphere_folder" "test" {
  path          = "foobar"
  type          = "vm"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"
}
```

Adding tag items triggers things appropriately:

```
* vsphere_tag_category.category: could not create category: Status code: 403: call failed: Error response from vCloud Suite API: {"type":"com.vmware.vapi.std.errors.unauthorized","value":{"messages":[]}}
```

Tag config was below:

```hcl
data "vsphere_datacenter" "dc" {
  name = "dc1"
}

resource "vsphere_tag_category" "category" {
  name        = "terraform-test-category"
  cardinality = "SINGLE"
  description = "Managed by Terraform"

  associable_types = [
    "Folder",
  ]
}

resource "vsphere_tag" "tag" {
  name        = "terraform-test-tag"
  category_id = "${vsphere_tag_category.category.id}"
  description = "Managed by Terraform"
}

resource "vsphere_folder" "test" {
  path          = "foobar"
  type          = "vm"
  datacenter_id = "${data.vsphere_datacenter.dc.id}"

  tags = ["${vsphere_tag.tag.id}"]
}
```

**Comment by deberon** [comment]
@vancluever I just tried it with 0.4.1 and I got the same result. I noticed this originally because my jobs were suddenly failing for any new terraform instances but one of the instances I instantiated a few weeks ago was still operational.

At the moment I am using terraform v0.10.7. I have also tried terraform v0.10.6 and v0.10.3. During troubleshooting I created a user with the perms outlined at https://www.terraform.io/docs/providers/vsphere/index.html#required-privileges and used this for my testing.

**Comment by vancluever** [comment]
Hey @deberon I still could not reproduce this, but it seems like vSphere does not necessarily like constantly messing around with permissions anyway - sessions seem to be retained over permission modification, which has made testing this a bit rough.

I do see in the code where this would be happening though and I don't know if there's much we can really do about it. The behaviour that we undertake is to not use the tags client for create and update calls if there are no pending tag update operations (ie: nothing defined in config or need to remove tags that were defined OOB), but on _read_ calls we need to read tags if we have a connected client to get the current state of tags on a resource regardless if they are defined in config or not (to account for that OOB situation again).

I don't really see that going away, hence I think that tagging privileges will be necessary on the permission list. So I will be re-opening your doc PR and at the very least we will make a note about the tagging support when we update the docs for v1.0.0.

Just before I close this issue out - adding the tag permissions fixed the read/refresh issues for you, yes?

Thanks!

**Comment by deberon** [comment]
@vancluever sounds good. And yeah, for my use case adding those permissions fixed the issue. Thanks for your prompt responses!

**Comment by deberon** [comment]
@vancluever you said you weren't able to reproduce this but you also see where in the code this would be happening correct? I seem to be having a weird problem now where I get the error when creating a sub folder, but not the parent folder. Output below. Could the code you saw be causing this as well? I'm beginning to think there is something screwed up with my vcenter server.

```
$ terraform apply
data.vsphere_datacenter.datacenter: Refreshing state...
vsphere_folder.folder: Creating...
  datacenter_id: "" => "datacenter-21"
  path:          "" => "folder_test"
  type:          "" => "vm"
vsphere_folder.folder: Creation complete after 0s (ID: group-v3356)
vsphere_folder.folder_sub: Creating...
  datacenter_id: "" => "datacenter-21"
  path:          "" => "folder_test/sub_test"
  type:          "" => "vm"
Error applying plan:

1 error(s) occurred:

* vsphere_folder.folder_sub: 1 error(s) occurred:

* vsphere_folder.folder_sub: error reading tags: Get unexpected status code: 403: call failed: Error response from vCloud Suite API: {"value":{"messages":[]},"type":"com.vmware.vapi.std.errors.unauthorized"}

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

**Comment by vancluever** [comment]
Hey @deberon, yeah, sorry, you are correct. `Read` calls happen on create and update too, so technically this call is happening on _all_ stages of a Terraform run except for destroy.

One thing I did notice during my troubleshooting of the issue is that there was a number of vSphere sessions still open that were interfering with my ability to effectively get new permissions after I assigned them. `govc`'s `session.ls` and `session.rm` commands were useful in flushing all of the relevant sessions. Not too sure if this will help you, but it's worth a shot (but remember that as mentioned tagging privileges will be necessary).
